sycl/include/CL/sycl/intel/builtins.hpp

```diff
@@ -0,0 +1,70 @@
+//==------ builtins.hpp - Non-standard SYCL built-in functions -------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+
+#ifdef __SYCL_DEVICE_ONLY__
+#define CONSTANT_AS __attribute__((opencl_constant))
+// Note: __format string is declared in constant address space to be compatible
+// with OpenCL C
+extern int __spirv_ocl_printf(const CONSTANT_AS char *__format, ...);
+#else
+#define CONSTANT_AS
+#endif
+
+namespace cl {
+namespace sycl {
+namespace intel {
+namespace experimental {
+
+// Provides functionality to print data from kernels in a C way:
+// - On non-HOST devices this function is directly mapped to printf from
+//   OpenCL C
+// - On host device, this function should be equivalent to standard printf
+//   function from C/C++.
+//
+// Please refer to corresponding section in OpenCL C specification to find
+// information about format string and its differences from standard C rules.
+//
+// This function is placed under 'experimental' namespace on purpose, because it
+// has too much caveats you need to be aware of before using it. Please find
+// them below and read carefully before using it:
+//
+// - According to the OpenCL spec, the format is must be
+// resolvable at compile time i.e. cannot be dynamically created by the
+// executing program, itself.
+//
+// - According to the OpenCL spec, the format string must reside in constant
+// address space. This requires to perform "tricky" declarations of them, see
+// test/built-ins/printf.cpp for examples
+// FIXME: this potentially can be done on SYCL FE side automatically
+//
+// - The format string is interpreted according to the OpenCL C spec, where all
+// data types has fixed bitwidth, opposed to C++ types which doesn't guarantee
+// the exact width of particular data types (except, may be, char). This might
```

Comment:
"except, may be, char": not really! This is why there is `CHAR_BITS` macro...

Reply:
> "except, may be, char": not really!

I guess it depends on units used to measure size: `sizeof` returns size in bytes and it is guaranteed that `sizeof(char) == 1`. So, in terms of bytes "except, may be, char" is a correct statement. In terms of bits, apparently no, because of `CHAR_BITS`

Reply:
Although you refer to "bitwidth", not bytes. Maybe just rename it to "size"?

Reply:
Renamed to "size"